File: legacy/packages/lsp/src/symbol-index.ts
import type { Program, Span, TypeSpec } from "@moon/ast";
import { existsSync, readdirSync, readFileSync } from "fs";
import { basename, join, resolve } from "path";
import { parse } from "@moon/parser";
import { resolveStdlibPath } from "@moon/resolver";
import { buildPrelude, formatType } from "@moon/typechecker";
import { allCoreModulePaths, coreModuleSchemes } from "@moon/typechecker/stdlib";
import { freshVar, instantiate, type Scheme, typeSpecToScheme } from "@moon/typechecker/types";
import { extractMoonDocs } from "./moon-docs";

export type SymbolKind =
  | "function"
  | "agent"
  | "model"
  | "data"
  | "constructor"
  | "type"
  | "value";

export interface SymbolRange {
  start: { line: number; character: number };
  end: { line: number; character: number };
}

export interface SymbolEntry {
  name: string;
  kind: SymbolKind;
  module: string;
  type: string;
  file: string;
  range: SymbolRange;
  docs?: string;
}

export interface MoonSymbolsFile {
  version: 1;
  updatedAt: string;
  projectRoot: string;
  symbols: SymbolEntry[];
}

function formatScheme(scheme: Scheme): string {
  return formatType(instantiate(scheme, { fresh: freshVar }));
}

function formatTypeSpec(spec: TypeSpec): string {
  return formatScheme(typeSpecToScheme(spec));
}

export function spanToRange(span: Span, name: string): SymbolRange {
  const line = Math.max(0, span.start.line - 1);
  const character = Math.max(0, span.start.column - 1);
  return {
    start: { line, character },
    end: { line, character: character + name.length },
  };
}

function schemeType(schemes: Map<string, Scheme> | null | undefined, name: string): string {
  const scheme = schemes?.get(name);
  return scheme ? formatScheme(scheme) : "?";
}

export function indexProgram(
  program: Program,
  source: string,
  filePath: string,
  moduleName: string,
  schemes?: Map<string, Scheme>,
): SymbolEntry[] {
  const abs = resolve(filePath);
  const entries: SymbolEntry[] = [];

  const push = (
    name: string,
    kind: SymbolKind,
    span: Span,
    type: string,
    docs?: string,
  ) => {
    entries.push({
      name,
      kind,
      module: moduleName,
      type,
      file: abs,
      range: spanToRange(span, name),
      docs: docs ?? extractMoonDocs(source, span.start.line),
    });
  };

  for (const decl of program.declarations) {
    switch (decl.kind) {
      case "Model":
        push(
          decl.decl.name,
          "model",
          decl.span,
          `model ${decl.decl.name}`,
        );
        for (const field of decl.decl.fields) {
          push(field.name, "type", field.span, formatTypeSpec(field.type));
        }
        break;
      case "Agent":
        push(
          decl.decl.name,
          "agent",
          decl.span,
          schemeType(schemes, decl.decl.name) || `agent ${decl.decl.name}`,
        );
        break;
      case "Data":
        push(decl.decl.name, "data", decl.span, `data ${decl.decl.name}`);
        for (const ctor of decl.decl.constructors) {
          push(ctor.name, "constructor", ctor.span, ctor.name);
        }
        break;
      case "Function":
        if (decl.decl.signature) {
          push(
            decl.decl.signature.name,
            "function",
            decl.decl.signature.span,
            formatTypeSpec(decl.decl.signature.type),
          );
        }
        for (const eq of decl.decl.equations) {
          if (decl.decl.signature?.name === eq.name) continue;
          push(
            eq.name,
            "function",
            eq.span,
            schemeType(schemes, eq.name),
          );
        }
        break;
      default:
        break;
    }
  }

  return entries;
}

function findNameLine(lines: string[], name: string): number {
  const patterns = [
    new RegExp(`^${name}\\s*::`),
    new RegExp(`^${name}\\s*=`),
    new RegExp(`^data\\s+${name}\\b`),
    new RegExp(`^agent\\s+${name}\\b`),
    new RegExp(`^model\\s+${name}\\b`),
    new RegExp(`=\\s*${name}\\b`),
  ];
  for (let i = 0; i < lines.length; i++) {
    const trimmed = lines[i]!.trim();
    if (patterns.some((re) => re.test(trimmed))) return i;
  }
  return -1;
}

function indexFromSourceText(
  source: string,
  filePath: string,
  moduleName: string,
): SymbolEntry[] {
  const schemes = coreModuleSchemes(moduleName);
  if (!schemes) return [];

  const abs = resolve(filePath);
  const lines = source.split(/\r?\n/);
  const entries: SymbolEntry[] = [];

  for (const [name, scheme] of schemes) {
    const lineIdx = findNameLine(lines, name);
    const line = Math.max(0, lineIdx);
    const raw = lineIdx >= 0 ? lines[lineIdx]! : "";
    const col = Math.max(0, raw.indexOf(name));
    entries.push({
      name,
      kind: "function",
      module: moduleName,
      type: formatScheme(scheme),
      file: abs,
      range: {
        start: { line, character: col },
        end: { line, character: col + name.length },
      },
      docs: lineIdx >= 0 ? extractMoonDocs(source, lineIdx + 1) : undefined,
    });
  }

  for (let i = 0; i < lines.length; i++) {
    const dataMatch = lines[i]!.trim().match(/^data\s+(\w+)\s*=/);
    if (!dataMatch) continue;
    const typeName = dataMatch[1]!;
    if (entries.some((e) => e.name === typeName)) continue;
    entries.push({
      name: typeName,
      kind: "data",
      module: moduleName,
      type: `data ${typeName}`,
      file: abs,
      range: spanToRange(
        { start: { line: i + 1, column: lines[i]!.indexOf(typeName) + 1, offset: 0 }, end: { line: i + 1, column: lines[i]!.indexOf(typeName) + 1 + typeName.length, offset: 0 } },
        typeName,
      ),
    });
    const ctorMatch = lines[i]!.match(/=\s*(\w+)/);
    if (ctorMatch && !entries.some((e) => e.name === ctorMatch[1])) {
      const ctor = ctorMatch[1]!;
      entries.push({
        name: ctor,
        kind: "constructor",
        module: moduleName,
        type: ctor,
        file: abs,
        range: spanToRange(
          { start: { line: i + 1, column: lines[i]!.indexOf(ctor) + 1, offset: 0 }, end: { line: i + 1, column: lines[i]!.indexOf(ctor) + 1 + ctor.length, offset: 0 } },
          ctor,
        ),
      });
    }
  }

  return entries;
}

function indexMoonFile(filePath: string, moduleName: string): SymbolEntry[] {
  const source = readFileSync(filePath, "utf-8");
  const schemes = coreModuleSchemes(moduleName) ?? undefined;
  try {
    const program = parse(source);
    return indexProgram(program, source, filePath, moduleName, schemes);
  } catch {
    return indexFromSourceText(source, filePath, moduleName);
  }
}

export function indexStdlib(): SymbolEntry[] {
  const entries: SymbolEntry[] = [];
  for (const modulePath of allCoreModulePaths()) {
    const stdlibPath = resolveStdlibPath(modulePath.split("."));
    if (!stdlibPath || !existsSync(stdlibPath)) continue;
    entries.push(...indexMoonFile(stdlibPath, modulePath));
  }
  return entries;
}

export function indexPrelude(): SymbolEntry[] {
  const prelude = buildPrelude();
  const entries: SymbolEntry[] = [];
  for (const [name, scheme] of prelude.values) {
    entries.push({
      name,
      kind: "type",
      module: "Moon.Prelude",
      type: formatScheme(scheme),
      file: "",
      range: { start: { line: 0, character: 0 }, end: { line: 0, character: name.length } },
      docs: "Moon prelude builtin",
    });
  }
  return entries;
}

export function indexWorkspace(projectRoot: string): SymbolEntry[] {
  const entries: SymbolEntry[] = [];
  const libDir = join(projectRoot, "lib");
  if (!existsSync(libDir)) return entries;

  for (const file of readdirSync(libDir).filter((f) => f.endsWith(".moon"))) {
    const filePath = join(libDir, file);
    const moduleName = basename(file, ".moon");
    entries.push(...indexMoonFile(filePath, moduleName));
  }
  return entries;
}

export function buildSymbolIndex(projectRoot: string, extraFiles: string[] = []): SymbolEntry[] {
  const seen = new Set<string>();
  const out: SymbolEntry[] = [];

  const add = (entry: SymbolEntry) => {
    const key = `${entry.module}::${entry.name}::${entry.file}`;
    if (seen.has(key)) return;
    seen.add(key);
    out.push(entry);
  };

  for (const entry of indexPrelude()) add(entry);
  for (const entry of indexStdlib()) add(entry);
  for (const entry of indexWorkspace(projectRoot)) add(entry);

  for (const filePath of extraFiles) {
    if (!existsSync(filePath)) continue;
    const source = readFileSync(filePath, "utf-8");
    const program = parse(source);
    const moduleName = basename(filePath, ".moon");
    for (const entry of indexProgram(program, source, filePath, moduleName)) add(entry);
  }

  return out.sort((a, b) => a.name.localeCompare(b.name) || a.module.localeCompare(b.module));
}

File: Sources/MoonLSP/SymbolIndex.swift
import Foundation
import MoonAST
import MoonMoonfile
import MoonParser
import MoonResolver
import MoonTypechecker
import MoonTypes

public enum SymbolKind: String, Codable, Sendable {
    case function
    case agent
    case model
    case data
    case constructor
    case type
    case value
}

public struct SymbolRange: Codable, Sendable, Equatable {
    public var start: Position
    public var end: Position

    public struct Position: Codable, Sendable, Equatable {
        public var line: Int
        public var character: Int
    }
}

public struct SymbolEntry: Codable, Sendable, Equatable {
    public var name: String
    public var kind: SymbolKind
    public var module: String
    public var type: String
    public var file: String
    public var range: SymbolRange
    public var docs: String?
}

public struct MoonSymbolsFile: Codable, Sendable, Equatable {
    public var version: Int
    public var updatedAt: String
    public var projectRoot: String
    public var symbols: [SymbolEntry]
}

public func spanToRange(_ span: Span, name: String) -> SymbolRange {
    let line = max(0, span.start.line - 1)
    let character = max(0, span.start.column - 1)
    return SymbolRange(
        start: .init(line: line, character: character),
        end: .init(line: line, character: character + name.count)
    )
}

public func indexProgram(
    _ program: Program,
    source: String,
    filePath: String,
    moduleName: String
) -> [SymbolEntry] {
    let abs = URL(fileURLWithPath: filePath).standardizedFileURL.path
    var entries: [SymbolEntry] = []

    func push(_ name: String, _ kind: SymbolKind, _ span: Span, _ type: String, _ docs: String? = nil) {
        entries.append(SymbolEntry(
            name: name,
            kind: kind,
            module: moduleName,
            type: type,
            file: abs,
            range: spanToRange(span, name: name),
            docs: docs ?? extractMoonDocs(source, declLine: span.start.line)
        ))
    }

    for decl in program.declarations {
        switch decl {
        case .model(let m, let span):
            push(m.name, .model, span, "model \(m.name)")
            for field in m.fields {
                push(field.name, .type, field.span, formatType(instantiate(typeSpecToScheme(field.type), supply: freshVar)))
            }
        case .agent(let a, let span):
            push(a.name, .agent, span, "agent \(a.name)")
        case .data(let d, let span):
            push(d.name, .data, span, "data \(d.name)")
            for ctor in d.constructors {
                push(ctor.name, .constructor, ctor.span, ctor.name)
            }
        case .function(let f, _):
            if let sig = f.signature {
                push(sig.name, .function, sig.span, formatType(instantiate(typeSpecToScheme(sig.type), supply: freshVar)))
            }
            for eq in f.equations {
                if f.signature?.name == eq.name { continue }
                push(eq.name, .function, eq.span, eq.name)
            }
        default:
            break
        }
    }
    return entries
}

public func indexStdlibSymbols() -> [SymbolEntry] {
    var entries: [SymbolEntry] = []
    for modulePath in allCoreModulePaths() {
        let parts = modulePath.split(separator: ".").map(String.init)
        guard let stdlibPath = resolveStdlibPath(modulePath: parts),
              let source = try? String(contentsOfFile: stdlibPath, encoding: .utf8) else { continue }
        if let program = try? MoonParser().parse(source) {
            entries.append(contentsOf: indexProgram(program, source: source, filePath: stdlibPath, moduleName: modulePath))
        }
    }
    return entries
}

public func indexPreludeSymbols() -> [SymbolEntry] {
    let prelude = buildPrelude()
    return prelude.values.map { name, scheme in
        SymbolEntry(
            name: name,
            kind: .type,
            module: "Moon.Prelude",
            type: formatType(instantiate(scheme, supply: freshVar)),
            file: "",
            range: SymbolRange(start: .init(line: 0, character: 0), end: .init(line: 0, character: name.count)),
            docs: "Moon prelude builtin"
        )
    }
}

public func indexWorkspaceSymbols(projectRoot: String) -> [SymbolEntry] {
    let libDir = URL(fileURLWithPath: projectRoot).appendingPathComponent("lib")
    guard let files = try? FileManager.default.contentsOfDirectory(atPath: libDir.path) else { return [] }
    var entries: [SymbolEntry] = []
    for file in files where file.hasSuffix(".moon") {
        let path = libDir.appendingPathComponent(file).path
        guard let source = try? String(contentsOfFile: path, encoding: .utf8),
              let program = try? MoonParser().parse(source) else { continue }
        let moduleName = String(file.dropLast(5))
        entries.append(contentsOf: indexProgram(program, source: source, filePath: path, moduleName: moduleName))
    }
    return entries
}

public func buildSymbolIndex(projectRoot: String, extraFiles: [String] = []) -> [SymbolEntry] {
    var seen = Set<String>()
    var out: [SymbolEntry] = []

    func add(_ entry: SymbolEntry) {
        let key = "\(entry.module)::\(entry.name)::\(entry.file)"
        guard !seen.contains(key) else { return }
        seen.insert(key)
        out.append(entry)
    }

    for entry in indexPreludeSymbols() { add(entry) }
    for entry in indexStdlibSymbols() { add(entry) }
    for entry in indexWorkspaceSymbols(projectRoot: projectRoot) { add(entry) }

    for filePath in extraFiles {
        guard FileManager.default.fileExists(atPath: filePath),
              let source = try? String(contentsOfFile: filePath, encoding: .utf8),
              let program = try? MoonParser().parse(source) else { continue }
        let moduleName = URL(fileURLWithPath: filePath).deletingPathExtension().lastPathComponent
        for entry in indexProgram(program, source: source, filePath: filePath, moduleName: moduleName) {
            add(entry)
        }
    }

    return out.sorted { a, b in
        let byName = a.name.localizedCaseInsensitiveCompare(b.name)
        if byName != .orderedSame { return byName == .orderedAscending }
        return a.module.localizedCaseInsensitiveCompare(b.module) == .orderedAscending
    }
}

public func findProjectRootFromPath(_ entryPath: String) -> String {
    var dir = URL(fileURLWithPath: entryPath).standardizedFileURL
    if dir.pathExtension == "moon" {
        dir = dir.deletingLastPathComponent()
    }
    while true {
        if findMoonfile(startDir: dir.path) != nil { return dir.path }
        let parent = dir.deletingLastPathComponent()
        if parent.path == dir.path { break }
        dir = parent
    }
    return URL(fileURLWithPath: entryPath).deletingLastPathComponent().path
}